tidy3d/plugins/smatrix/smatrix.py

```diff
@@ -69,50 +137,270 @@ class SMatrixDataArray(DataArray):
+    def batch_path(self) -> str:
+        """Path to the batch saved to file."""
+
+        return self.batch._batch_path(path_dir=DEFAULT_DATA_DIR)
+
+    def get_path_dir(self, path_dir: str) -> None:
+        """Check whether the supplied 'path_dir' matches the internal field value."""
+
+        if path_dir not in (DEFAULT_DATA_DIR, self.path_dir):
+            log.warning(
+                f"'ComponentModeler' method was supplied a 'path_dir' of '{path_dir}' "
+                f"when its internal 'path_dir' field was set to '{self.path_dir}'. "
+                "The passed value will be deprecated in later versions. "
+                "Please set the internal 'path_dir' field to the desired value and "
+                "remove the 'path_dir' from the method argument. "
+                f"Using supplied '{path_dir}'."
+            )
+            return path_dir
+
+        return self.path_dir
+
+    @cached_property
+    def _batch_path(self) -> str:
+        """Where we store the batch for this ComponentModeler instance after the run."""
+        return os.path.join(self.path_dir, "batch" + str(hash(self)) + ".json")
+
+    def _run_sims(self, path_dir: str = DEFAULT_DATA_DIR) -> BatchData:
+        """Run :class:`Simulations` for each port and return the batch after saving."""
+        batch = self.batch
+
+        # TEMP
+        # import sys
+        # TIDY3D_CORE_PATH = '/home/shashwat/flexcompute/repositories/tidy3d-core'
+        # sys.path.insert(0, TIDY3D_CORE_PATH)
+        # from tidy3d_backend.run import run_sim
+        # batch_data = {}
+        # for name, sim in self.sim_dict.items():
+        #     batch_data[name] = run_sim(sim, mpi=1)
+
+        # TEMP
+        batch_data = batch.run(path_dir=path_dir)
+
+        batch.to_file(self._batch_path)
+        return batch_data
+
+    def get_port_by_name(self, port_name: str) -> Port:
+        """Get the port from the name."""
+        ports = [port for port in self.ports if port.name == port_name]
+        if len(ports) == 0:
+            raise Tidy3dKeyError(f'Port "{port_name}" not found.')
+        return ports[0]
+
+    @abstractmethod
+    def _construct_smatrix(self, batch_data: BatchData) -> DataArray:
+        """Post process `BatchData` to generate scattering matrix."""
+
+    def run(self, path_dir: str = DEFAULT_DATA_DIR) -> DataArray:
+        """Solves for the scattering matrix of the system."""
+        path_dir = self.get_path_dir(path_dir)
+
+        batch_data = self._run_sims(path_dir=path_dir)
+        return self._construct_smatrix(batch_data=batch_data)
+
+    def load(self, path_dir: str = DEFAULT_DATA_DIR) -> DataArray:
+        """Load a scattering matrix from saved `BatchData` object."""
+        path_dir = self.get_path_dir(path_dir)
+
+        batch_data = BatchData.load(path_dir=path_dir)
+        return self._construct_smatrix(batch_data=batch_data)
+
+    @staticmethod
+    def s_to_z(s_matrix: DataArray, reference: Complex) -> DataArray:
```

Comment:
let's make sure these operations are tested in the unit tests (if not already). Like that they give expected matrices given some example inputs.